### PR TITLE
Return copy of deposits instead of internal pointer

### PR DESCRIPTION
**Pull request by jtraglia**
**What type of PR is this?**

Bug fix

**What does this PR do? Why is it needed?**

Just to be safe, I think we should return a copy of the deposits instead of the `DepositCache.deposits` pointer. There is a read lock in this method, so it doesn't make sense to return a pointer that can be read (and written to) outside of the lock. This could lead to a race condition if another thread modifies the deposits.

**Other notes for review**

All of the other methods for this interface return a copy.

### CODE REVIEW

**Comment by prestonvanloon** [comment]
What kind of memory or performance analysis have you done with this change? 

Copy on read can be expensive depending on the call frequency and cache size.

**Comment by jtraglia** [comment]
Hey @prestonvanloon good point. It seems that this method (`AllDepositContainers`) is only called in two places:

* [`ensureValidPowchainData`](https://github.com/prysmaticlabs/prysm/blob/2728b83f6aa7a78b2d014eb59e9428c58fd85a1b/beacon-chain/execution/service.go#L786) - seems to only be called once during startup.
* [`savePowchainData`](https://github.com/prysmaticlabs/prysm/blob/d077483577bc8fdc658940adf71d8d11dfa5949b/beacon-chain/execution/log_processing.go#L543) - will be called every so often.

Quick napkin math tells me that the copy would take up an additional ~4MB of memory.

> 500,000 deposits * 8 bytes (64-bit pointer) == 4MB

And for me, it would take approximately 0.22ms to perform the copy. [Some benchmarks here](https://gist.github.com/jtraglia/0cc09c5df3773cd4accffdba27d0c168).

But this isn't a hill that I'm willing to die on, especially being so close to the merge. It's just something I came across when writing some semgrep rules. There's already some level of inherent trust in Prysm. For example, this promise:

https://github.com/prysmaticlabs/prysm/blob/d077483577bc8fdc658940adf71d8d11dfa5949b/beacon-chain/execution/log_processing.go#L557

So yeah, it's up to you. I don't mind if we scrap this PR. It's not really a problem.

**Comment by prestonvanloon** [comment]
Deposits are rather small in comparison to the beacon state. I agree with your performance assessment, it is a small sacrifice for safer cache. If only golang had immutable objects...

@nisdas could you take a look at this before? Does this feel familiar?

**Comment by prestonvanloon** [comment]
> 500,000 deposits * 8 bytes (64-bit pointer) == 4MB

OK, wait a sec. I just re-read this and I don't think I agree. You must do a deep copy or downstream can still mutate the slices which are shared between structs. So if we are assuming a deep copy, each object is far more than 8 bytes

```go
// Total size for DepositContainer is 1256
type DepositContainer struct {
	Index           int64    // 8 bytes
	Eth1BlockHeight uint64 // 8 bytes
	Deposit         *Deposit // 1208 bytes
	DepositRoot     []byte // 32 bytes
}

// Total size for Deposit is 1208 bytes
type Deposit struct {
	Proof [][]byte     // 32 * 32 bytes
	Data  *Deposit_Data  // 184 bytes
}

// Total size for Deposit_Data = 184 bytes
type Deposit_Data struct {
	PublicKey             []byte // 48 bytes
	WithdrawalCredentials []byte // 32 bytes
	Amount                uint64  // 8 bytes
	Signature             []byte  // 96 bytes
}
```

Assuming the above is correct and each byte slice is a distinct array of the exact size, then the math is more like this:

500,000 deposits * (1256 + 3 * 8 bytes (64-bit pointer)) == ~610Mi

**Comment by jtraglia** [comment]
Yeah, I wasn't requesting a deep copy, just a copy of the list of deposits (`[]*ethpb.DepositContainer`). Sure, it's possible to change an individual deposit, but that's much less likely than an additional deposit showing up in the list later, maybe while it's being iterated over or something. I agree that copying the whole thing doesn't make sense.

**Comment by jtraglia** [comment]
Just to be clear, it returns a list of deposit (`ethpb.DepositContainer`) pointers which would be at most 8 bytes each.

**Comment by jtraglia** [comment]
That comment I put above the copy is quite misleading now though. What do you want to do?

**Comment by jtraglia** [comment]
Yeah, just mentioned that a few seconds before. I agree.

**Comment by infosecual** [comment]
I stumbled on this PR while reviewing the `depositsLock` mutex and scratched my head for a good bit so I thought I would leave a comment in hopes that it helps someone else. At first I thought that all of the following functions were unsafe because they returned references to `Deposit` objects:

**Comment by infosecual** [comment]
I stumbled on this PR while reviewing the `depositsLock` mutex and scratched my head for a good bit so I thought I would leave a comment in hopes that it helps someone else. Initially I thought that all of the following functions were unsafe because they returned references to `Deposit` objects:
```
(dc *DepositCache) AllDepositContainers
(dc *DepositCache) allDeposits
(dc *DepositCache) NonFinalizedDeposits
(dc *DepositCache) PendingContainers
```
In reality I got confused about what the `depositsLock` mutex is really guarding. ref `DepositCache` struct:
```
// DepositCache stores all in-memory deposit objects. This
// stores all the deposit related data that is required by the beacon-node.
type DepositCache struct {
	// Beacon chain deposits in memory.
	pendingDeposits   []*ethpb.DepositContainer
	deposits          []*ethpb.DepositContainer
	finalizedDeposits *FinalizedDeposits
	depositsByKey     map[[fieldparams.BLSPubkeyLength]byte][]*ethpb.DepositContainer
	depositsLock      sync.RWMutex
}
```
It seems `depositsLock` guards the `pendingDeposits`  and `deposits` slices, but NOT the individual `DepositContainer` structs that those slices contain. I followed the lifetime of the individual `DepositContainer` objects and they do not have race issues because they are never written to after they are initialized and become readable. They are not destroyed either, just left to be collected by the GC (see `(dc *DepositCache) PrunePendingDeposits`).

In short- I think `(dc *DepositCache) allDeposits`, `(dc *DepositCache) NonFinalizedDeposits`, and `(dc *DepositCache) PendingContainers` are already safe since they make return new slices and @jtraglia's PR makes `(dc *DepositCache) AllDepositContainers` safe now.

Since the underlying `Deposit` objects are not racy I believe a shallow copy of the referencing slice is sufficient from a safety perspective. As for the performance hit, I'll defer to others. If it is determined that the performance hit is unacceptable I think we have some options to make more fine gained accessors since the caller's of these often only use one subcomponent of the returned deposits (eg just the deposits' root hashes), for which we can make a separate accessor with less overhead.

**Comment by nisdas** [comment]
@prestonvanloon This particular method is only ever used when saving execution related data to the db. This can lead to a slight increase in memory usage as you allocate a whole new slice for the pointers. ( 400000 * 8 = 3.2 mb) . I don't think it is too big a deal adding it in. If it was copying all the deposits, that would be a bigger issue. The only tricky thing would be that when processing a large amount of deposit logs in a period of time(syncing), you would have slightly higher memory usage for it. But the GC should be able to handle this easily, so mostly a non-issue.

**Review by prestonvanloon — APPROVED** [review]
